src/services/inlayHints.ts

```diff
@@ -0,0 +1,291 @@
+            }
+
+            const returnType = checker.getReturnTypeOfSignature(signature);
+            if (isModuleReferenceType(returnType)) {
+                return;
+            }
+
+            const typeDisplayString = printTypeInSingleLine(returnType);
+            if (!typeDisplayString) {
+                return;
+            }
+
+            addTypeHints(typeDisplayString, getTypeAnnotationPosition(decl));
+        }
+
+        function getTypeAnnotationPosition(decl: ArrowFunction | FunctionExpression | MethodDeclaration | FunctionDeclaration) {
+            const closeParenToken = findChildOfKind(decl, SyntaxKind.CloseParenToken, file);
+            if (closeParenToken) {
+                return closeParenToken.end;
+            }
+            return decl.parameters.end;
+        }
+
+        function visitFunctionExpressionLikeForParameterType(expr: ArrowFunction | FunctionExpression) {
+            if (!expr.parameters.length || expr.parameters.every(param => !!getEffectiveTypeAnnotationNode(param))) {
+                return;
+            }
+
+            const contextualType = checker.getContextualType(expr);
+            if (!contextualType) {
+                return;
+            }
+
+            const signatures = checker.getSignaturesOfType(contextualType, SignatureKind.Call);
+            const signature = firstOrUndefined(signatures);
+            if (!signature) {
+                return;
+            }
+
+            for (let i = 0; i < expr.parameters.length && i < signature.parameters.length; ++i) {
+                const param = expr.parameters[i];
+                const effectiveTypeAnnotation = getEffectiveTypeAnnotationNode(param);
+
+                if (effectiveTypeAnnotation) {
+                    continue;
+                }
+
+                const typeDisplayString = getParameterDeclarationTypeDisplayString(signature.parameters[i]);
+                if (!typeDisplayString) {
+                    continue;
+                }
+
+                addTypeHints(typeDisplayString, param.end);
+            }
+        }
+
+        function getParameterDeclarationTypeDisplayString(symbol: Symbol) {
+            const valueDeclaration = symbol.valueDeclaration;
+            if (!valueDeclaration || !isParameter(valueDeclaration)) {
+                return undefined;
+            }
+
+            const signatureParamType = checker.getTypeOfSymbolAtLocation(symbol, valueDeclaration);
+            if (isModuleReferenceType(signatureParamType)) {
+                return undefined;
+            }
+
+            return printTypeInSingleLine(signatureParamType);
+        }
+
+        function truncation(text: string, maxLength: number) {
+            if (text.length > maxLength) {
+                return text.substr(0, maxLength - "...".length) + "...";
+            }
+            return text;
+        }
+
+        function createSignleLineWriter(writer: DisplayPartsSymbolWriter): DisplayPartsSymbolWriter {

```

Comment:
```suggestion
        function createSingleLineWriter(writer: DisplayPartsSymbolWriter): DisplayPartsSymbolWriter {
```

Reply:
Surely there must be something like this already? @weswigham @andrewbranch

Reply:
It might be `createSingleLineStringWriter` in `src/compiler/utilities.ts`